src/backend/distributed/transaction/backend_data.c

```diff
@@ -780,6 +784,85 @@ MarkCitusInitiatedCoordinatorBackend(void)
 }
 
 
+/*
+ * AssignGlobalPID assigns a global process id for the current backend
+ */
+void
+AssignGlobalPID(void)
+{
+	if (MyBackendData->globalPID != 0)
+	{
+		elog(DEBUG1, "a global pid is already assigned to this backend");
+		return;
+	}
+
+	if (!IsCitusInitiatedRemoteBackend())
+	{
+		uint64 globalPID = GenerateGlobalPID();
+
+		SpinLockAcquire(&MyBackendData->mutex);
+		MyBackendData->globalPID = globalPID;
+		SpinLockRelease(&MyBackendData->mutex);
+
+		return;
+	}
+
+	uint64 globalPID = ExtractGlobalPID();
+
+	SpinLockAcquire(&MyBackendData->mutex);
+	MyBackendData->globalPID = globalPID;
+	SpinLockRelease(&MyBackendData->mutex);
+}
+
+
+/*
+ * GetGlobalPID returns the global process id of the current backend
+ */
+uint64
+GetGlobalPID(void)
+{
+	uint64 globalPID = 0;
+	SpinLockAcquire(&MyBackendData->mutex);
+	globalPID = MyBackendData->globalPID;
+	SpinLockRelease(&MyBackendData->mutex);
+	return globalPID;
+}
+
+
+/*
+ * GenerateGlobalPID generates the global process id for the current backend
+ */
+static uint64
+GenerateGlobalPID(void)
+{
+	/*
+	 * We try to create a human readable global pid that consists of node id and process id.
+	 * By multiplying node id with 10^10 and adding pid we generate a number where the smallest
+	 * 10 digit represent the pid and the remaining digits are the node id
+	 * 
+	 * Both node id and pid are 32 bit. We use 10^10 to fit all possible pids. Some very large
+	 * node ids might cause overflow. But even for the applications that scale around 50 nodes every
+	 * day it'd take about 100K years. So we are not worried.
+	 */
+	return (((uint64) GetLocalNodeId()) * 10000000000) + getpid();
+}
+
+/*
+ * ExtractGlobalPID extracts the global process id from the application name  
+ */
+static uint64
+ExtractGlobalPID(void)
+{
+	StringInfo applicationName = makeStringInfo();
```

Comment:
Shouldn't we call `AssignGlobalPID` to make sure it is assigned?

Reply:
Ok I do this and remove the message in the assigning function that says "gpid is already assigned"

Reply:
I realized that we shouldn't call `AssignGlobalPID`. Because this function doesn't actually look at the current backend's global pid. It extracts the gpid from the application name. So assigning won't do anything.